homeassistant/scripts/check_config.py

```diff
@@ -0,0 +1,172 @@
+    parser.add_argument(
+        '--script', choices=['check_config'])
+    parser.add_argument(
+        '--verbose', default=False)
+
+    args = parser.parse_args()
+
+    config_dir = os.path.join(os.getcwd(), args.config)
+    config_path = os.path.join(config_dir, 'configuration.yaml')
+    if not os.path.isfile(config_path):
+        print('Config does not exist:', config_path)
+        return 1
+
+    print(color("Testing configuration at", config_dir))
+
+    res = check(config_path)
+
+    if SHOW_FILES:  # Show Files
+        print(color('yaml files'), '(used /',
+              color('not used', c=Fore.RED)+')')
+        # Python 3.5 gets a recursive, but not in 3.4
+        for yfn in sorted(glob(os.path.join(config_dir, '*.yaml')) +
+                          glob(os.path.join(config_dir, '*/*.yaml'))):
+            the_color = '' if yfn in res['yaml_files'] else Fore.RED
+            print(color('-', yfn, c=the_color))
+
+    # pylint: disable=too-many-nested-blocks
+    if SHOW_FULL_CONFIG:
+        re_keys = re.compile(r"'([^']+)':")
+
+        def indent(item, indent0, indent1):
+            """Return indented text."""
+            return re_keys.sub(color(r'\1')+':',
+                               fill(pprint.pformat(item),
+                                    initial_indent=indent0,
+                                    subsequent_indent=indent1))
+
+        def printlist(obj, level=0, pref=''):
+            """Printlist."""
+            indent1 = level * ' '
+            indent0 = pref.rjust(level)
+            if isinstance(obj, list):
+                for ob0 in obj:
+                    printlist(ob0, level+2, '- ')
+            # elif isinstance(obj, dict):
+            #    for key0, val0 in obj.items():
+            #        printlist(val0, indent00 + key0 + ':' , indent1)
+            else:
+                print(indent(obj, indent0, indent1))
+
+        components = deep_convert_dict(res['components'])
+        for name, config in components:
+            if config is None:
+                continue
+            print(color(name + ':'))
+            printlist(config, 2, '')
+
+    return 0
+
+
+def check(config_path):
+    """Perform a check by mocking hass load functions."""
+    # List of components/platforms
+    components = []
+    yaml_files = {}
+    secrets = {}
+
+    original_get_component = loader.get_component
+    original_load_yaml = yaml.load_yaml
+
+    def mock_load_yaml(filename):
+        """Mock hass.util.load_yaml to save config files."""
+        yaml_files[filename] = True
+        return original_load_yaml(filename)
+
+    def mock_get_component(comp_name):
+        """Mock hass.loader.get_component to replace setup & setup_platform."""
+        def mock_setup(*kwargs):
+            """Mock setup, only record the component name & config."""
+            # pylint: disable=cell-var-from-loop
+            components.append((comp_name, kwargs[1].get(comp_name)))
```

Comment:
You're right. I was only thinking about getting the names. We might want to do both? btw, no need to name it `*kwargs` because a) it's a list, so `*args` and b) you know exactly what the input is going to be `mock_setup(hass, config)`, so just name it that.